### ISSUE TITLE
Computing confidence mask

### ISSUE THREAD

**Issue by alexandru-dinu** [eriklindernoren/PyTorch-YOLOv3#22]
In [`build_targets`](https://github.com/eriklindernoren/PyTorch-YOLOv3/blob/master/utils/utils.py#L130) function, at the beginning, there's a part that calculates the confidence mask tensor.
Initially it is set to a tensor of ones, but the update rule: 

        # Objects with higher confidence than threshold are set to zero
        conf_mask[b][cur_ious.view_as(conf_mask[b]) > ignore_thres] = 0

doesn't make sense to me. This basically ignores any ious better than `ignore_thres` (currently set to `0.5`). 

I'd think that:
- either start with a tensor of zeros and use the update rule: 
`conf_mask[b][cur_ious.view_as(conf_mask[b]) > thres] = 1`
- either change the sign: ` conf_mask[b][cur_ious.view_as(conf_mask[b]) <= thres] = 0`

Thanks

**Comment by PiggyGenius** [comment]
I found the part you are refering to in the commit history but I am not sure why this filtering was removed. Can you explain @eriklindernoren ?

**Comment by alexandru-dinu** [comment]
I guess because, it was redundant (or not needed all). The mask is computed as: 
`mask[b, best_anchor_idx, gj, gi] = 1`, basically saying that for batch `b`, the anchor `best_anchor_idx` of cell `(gj, gi)` is responsible for predicting some ground truth object.

**Comment by eriklindernoren** [comment]
Hi. This should be fixed in 959e0ff. Before this commit the confidence mask was computed as in YOLOv2. With the latest commit the confidence mask is set as in version 3, i.e. where the anchor has an overlap with a target box larger than 0.5 but is not the anchor with the largest overlap the confidence prediction is ignored during training.

**Comment by ydixon** [comment]
I found that `conf_mask[b, anch_ious > ignore_thres] = 0` will overwrite any previous ground truth target that has `conf_mask` set to 1.

        for t in range(target.shape[1]):
            if target[b, t].sum() == 0:
                continue
            .......
            # Calculate iou between gt and anchor shapes
            anch_ious = bbox_iou(gt_box, anchor_shapes)
            # Where the overlap is larger than threshold set mask to zero (ignore)
            conf_mask[b, anch_ious > ignore_thres] = 0    <---------------------
            # Find the best matching anchor box
            best_n = np.argmax(anch_ious)
            ...
            # Masks
            mask[b, best_n, gj, gi] = 1
            conf_mask[b, best_n, gj, gi] = 1    <-------------------

Let's say: 
```
Iter1: anch_ious = [0.567 0.305 0.43], anch_ious>ignore_thres=0,best_n=0, conf_mask[b, 0, gj, gi]= 1
Iter2: anch_ious = [0.667 0.045 0.33], anch_ious>ignore_thres=0,best_n=0, conf_mask[b, 0, gj, gi]= 1
Iter3: anch_ious = [0.734 0.025 0.22], anch_ious>ignore_thres=0,best_n=0, conf_mask[b, 0, gj, gi]= 1
```

So every iteration is erasing the previous iteration's ground truth target's `conf_mask` if the target is selecting the same anchor. For the example above, only `Iter3: conf_mask[b, 0, gj, gi]= 1`is kept. It seems like more than half of the ground truth target's are usually ignored during training. Is this intended behavior and why would it work for training?
